File: plugins/org.eclipse.tm.tcf/src/org/eclipse/tm/internal/tcf/services/remote/BreakpointsProxy.java
/*******************************************************************************
 * Copyright (c) 2007, 2008 Wind River Systems, Inc. and others.
 * All rights reserved. This program and the accompanying materials 
 * are made available under the terms of the Eclipse Public License v1.0 
 * which accompanies this distribution, and is available at 
 * http://www.eclipse.org/legal/epl-v10.html 
 *  
 * Contributors:
 *     Wind River Systems - initial API and implementation
 *******************************************************************************/
package org.eclipse.tm.internal.tcf.services.remote;

import java.io.IOException;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import org.eclipse.tm.tcf.core.Command;
import org.eclipse.tm.tcf.protocol.IChannel;
import org.eclipse.tm.tcf.protocol.IToken;
import org.eclipse.tm.tcf.protocol.JSON;
import org.eclipse.tm.tcf.services.IBreakpoints;


public class BreakpointsProxy implements IBreakpoints {

    private final IChannel channel;
    private final Map<BreakpointsListener,IChannel.IEventListener> listeners =
        new HashMap<BreakpointsListener,IChannel.IEventListener>();

    public BreakpointsProxy(IChannel channel) {
        this.channel = channel;
    }

    public IToken set(Map<String,Object>[] properties, final DoneCommand done) {
        return new Command(channel, this, "set", new Object[]{ properties }) {
            @Override
            public void done(Exception error, Object[] args) {
                if (error == null) {
                    assert args.length == 1;
                    error = toError(args[0]);
                }
                done.doneCommand(token, error);
            }
        }.token;
    }

    public IToken add(Map<String,Object> properties, final DoneCommand done) {
        return new Command(channel, this, "add", new Object[]{ properties }) {
            @Override
            public void done(Exception error, Object[] args) {
                if (error == null) {
                    assert args.length == 1;
                    error = toError(args[0]);
                }
                done.doneCommand(token, error);
            }
        }.token;
    }

    public IToken change(Map<String,Object> properties, final DoneCommand done) {
        return new Command(channel, this, "change", new Object[]{ properties }) {
            @Override
            public void done(Exception error, Object[] args) {
                if (error == null) {
                    assert args.length == 1;
                    error = toError(args[0]);
                }
                done.doneCommand(token, error);
            }
        }.token;
    }

    public IToken disable(String[] ids, final DoneCommand done) {
        return new Command(channel, this, "disable", new Object[]{ ids }) {
            @Override
            public void done(Exception error, Object[] args) {
                if (error == null) {
                    assert args.length == 1;
                    error = toError(args[0]);
                }
                done.doneCommand(token, error);
            }
        }.token;
    }

    public IToken enable(String[] ids, final DoneCommand done) {
        return new Command(channel, this, "enable", new Object[]{ ids }) {
            @Override
            public void done(Exception error, Object[] args) {
                if (error == null) {
                    assert args.length == 1;
                    error = toError(args[0]);
                }
                done.doneCommand(token, error);
            }
        }.token;
    }

    public IToken remove(String[] ids, final DoneCommand done) {
        return new Command(channel, this, "remove", new Object[]{ ids }) {
            @Override
            public void done(Exception error, Object[] args) {
                if (error == null) {
                    assert args.length == 1;
                    error = toError(args[0]);
                }
                done.doneCommand(token, error);
            }
        }.token;
    }

    public IToken getIDs(final DoneGetIDs done) {
        return new Command(channel, this, "getIDs", null) {
            @Override
            public void done(Exception error, Object[] args) {
                String[] arr = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    arr = toStringArray(args[1]);
                }
                done.doneGetIDs(token, error, arr);
            }
        }.token;
    }

    public IToken getProperties(String id, final DoneGetProperties done) {
        return new Command(channel, this, "getProperties", new Object[]{ id }) {
            @SuppressWarnings("unchecked")
            @Override
            public void done(Exception error, Object[] args) {
                Map<String,Object> map = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    map = (Map<String,Object>)args[1];
                }
                done.doneGetProperties(token, error, map);
            }
        }.token;
    }

    public IToken getStatus(String id, final DoneGetStatus done) {
        return new Command(channel, this, "getStatus", new Object[]{ id }) {
            @SuppressWarnings("unchecked")
            @Override
            public void done(Exception error, Object[] args) {
                Map<String,Object> map = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    map = (Map<String,Object>)args[1];
                }
                done.doneGetStatus(token, error, map);
            }
        }.token;
    }

    public IToken getCapabilities(String id, final DoneGetCapabilities done) {
        return new Command(channel, this, "getCapabilities", new Object[]{ id }) {
            @SuppressWarnings("unchecked")
            @Override
            public void done(Exception error, Object[] args) {
                Map<String,Object> map = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    map = (Map<String,Object>)args[1];
                }
                done.doneGetCapabilities(token, error, map);
            }
        }.token;
    }

    public String getName() {
        return NAME;
    }

    @SuppressWarnings("unchecked")
    private String[] toStringArray(Object o) {
        if (o == null) return null;
        Collection<String> c = (Collection<String>)o;
        return (String[])c.toArray(new String[c.size()]);
    }

    @SuppressWarnings("unchecked")
    private Map<String,Object>[] toBreakpointArray(Object o) {
        if (o == null) return null;
        Collection<Map<String,Object>> c = (Collection<Map<String,Object>>)o;
        return (Map<String,Object>[])c.toArray(new Map[c.size()]);
    }

    public void addListener(final BreakpointsListener listener) {
        IChannel.IEventListener l = new IChannel.IEventListener() {

            @SuppressWarnings("unchecked")
            public void event(String name, byte[] data) {
                try {
                    Object[] args = JSON.parseSequence(data);
                    if (name.equals("status")) {
                        assert args.length == 2;
                        listener.breakpointStatusChanged((String)args[0], (Map<String,Object>)args[1]);
                    }
                    else if (name.equals("contextAdded")) {
                        assert args.length == 1;
                        listener.contextAdded(toBreakpointArray(args[0]));
                    }
                    else if (name.equals("contextChanged")) {
                        assert args.length == 1;
                        listener.contextChanged(toBreakpointArray(args[0]));
                    }
                    else if (name.equals("contextRemoved")) {
                        assert args.length == 1;
                        listener.contextRemoved(toStringArray(args[0]));
                    }
                    else {
                        throw new IOException("Breakpoints service: unknown event: " + name);
                    }
                }
                catch (Throwable x) {
                    channel.terminate(x);
                }
            }
        };
        channel.addEventListener(this, l);
        listeners.put(listener, l);
    }

    public void removeListener(BreakpointsListener listener) {
        IChannel.IEventListener l = listeners.remove(listener);
        if (l != null) channel.removeEventListener(this, l);
    }
}

File: python/src/tcf/services/remote/BreakpointsProxy.py
# *******************************************************************************
# * Copyright (c) 2011 Wind River Systems, Inc. and others.
# * All rights reserved. This program and the accompanying materials
# * are made available under the terms of the Eclipse Public License v1.0
# * which accompanies this distribution, and is available at
# * http://www.eclipse.org/legal/epl-v10.html
# *
# * Contributors:
# *     Wind River Systems - initial API and implementation
# *******************************************************************************

import exceptions
from tcf import channel
from tcf.services import breakpoints
from tcf.channel.Command import Command

class BPCommand(Command):
    def __init__(self, service, cmd, cb, *args):
        super(BPCommand, self).__init__(service.channel, service, cmd, args)
        self.__cb = cb
    def done(self, error, args):
        if not error:
            assert len(args) == 1
            error = self.toError(args[0])
        self.__cb.doneCommand(self.token, error)


class ChannelEventListener(channel.EventListener):
    def __init__(self, service, listener):
        self.service = service
        self.listener = listener
    def event(self, name, data):
        try:
            args = channel.fromJSONSequence(data)
            if name == "status":
                assert len(args) == 2
                self.listener.breakpointStatusChanged(args[0], args[1])
            elif name == "contextAdded":
                assert len(args) == 1
                self.listener.contextAdded(args[0])
            elif name == "contextChanged":
                assert len(args) == 1
                self.listener.contextChanged(args[0])
            elif name == "contextRemoved":
                assert len(args) == 1
                self.listener.contextRemoved(args[0])
            else:
                raise IOError("Breakpoints service: unknown event: " + name);
        except exceptions.Exception as x:
            self.service.channel.terminate(x)

class BreakpointsProxy(breakpoints.BreakpointsService):
    def __init__(self, channel):
        self.channel = channel
        self.listeners = {}

    def set(self, properties, done):
        return BPCommand(self, "set", done, properties).token

    def add(self, properties, done):
        return BPCommand(self, "add", done, properties).token

    def change(self, properties, done):
        return BPCommand(self, "change", done, properties).token

    def disable(self, ids, done):
        return BPCommand(self, "disable", done, ids).token

    def enable(self, ids, done):
        return BPCommand(self, "enable", done, ids).token

    def remove(self, ids, done):
        return BPCommand(self, "remove", done, ids).token

    def getIDs(self, done):
        service = self
        class GetIDsCommand(Command):
            def __init__(self):
                super(GetIDsCommand, self).__init__(service.channel, service, "getIDs", None)
            def done(self, error, args):
                ids = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    ids = args[1]
                done.doneGetIDs(self.token, error, ids)
        return GetIDsCommand().token

    def getProperties(self, id, done):
        service = self
        class GetPropertiesCommand(Command):
            def __init__(self):
                super(GetPropertiesCommand, self).__init__(service.channel, service, "getProperties", (id,))
            def done(self, error, args):
                map = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    map = args[1]
                done.doneGetProperties(self.token, error, map)
        return GetPropertiesCommand().token

    def getStatus(self, id, done):
        service = self
        class GetStatusCommand(Command):
            def __init__(self):
                super(GetStatusCommand, self).__init__(service.channel, service, "getStatus", (id,))
            def done(self, error, args):
                map = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    map = args[1]
                done.doneGetStatus(self.token, error, map)
        return GetStatusCommand().token

    def getCapabilities(self, id, done):
        service = self
        class GetCapabilitiesCommand(Command):
            def __init__(self):
                super(GetCapabilitiesCommand, self).__init__(service.channel, service, "getCapabilities", (id,))
            def done(self, error, args):
                map = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    map = args[1]
                done.doneGetCapabilities(self.token, error, map)
        return GetCapabilitiesCommand().token

    def addListener(self, listener):
        l = ChannelEventListener(self, listener)
        self.channel.addEventListener(self, l)
        self.listeners[listener] = l

    def removeListener(self, listener):
        l = self.listeners.get(listener)
        if l:
            del self.listeners[listener]
            self.channel.removeEventListener(self, l)
